### DISCUSSION TITLE
Dynamic message keys

### DISCUSSION THREAD

**Discussion by henrikbossart** [amannn/next-intl#1607] [Q&A] [answered]
Hi,

I have am trying to get the mesages dynamically like this:

```typescript
// defined in messages.ts
export default {
  PageTitle: {
    new: {...},
    reports: {...},
    ...
  }
}


const titleTranslations = useTranslations('PageTitles')
const title =
    props.title ?? titleTranslations(`${pathname.split('/').at(-1) as typeof keys}.title`)
```

But am having a hard time doing it because TypeScript can't stop complaining about it. It works if I explicitly define the variable to be a certain key.

Is there some way I can type or cast this to work with dynamic keys?

Any help is highly appreciated! 🙏

**Comment by MaximeDetaille** [comment]
I actually got the same problem and i dont know how to fix that ...

**Comment by plfmoura-redraw** [comment]
Same problem here

**Comment by abdullah-cod9** [comment, marked as answer]
Try this

`type Theme = "dark" | "light" | "system";`
`{t('general.theme.${theme as Theme}')}`

en.json = 
`{
"theme": {
          "title": "Theme",
          "dark": "Dark",
          "light": "Light",
          "system": "System"
        },
}`
it worked with me

**Reply by amannn** [reply]
[`select`](https://next-intl.dev/docs/usage/messages#selecting-enum-based-values) can also work well for these cases:

```json
"theme": "{name, select, light {Light} dark {Dark} system {System} other {Unknown}}"
```

```tsx
t('general.theme', {name: 'light'});
```

**Reply by henrikbossart** [reply]
That works. I was reluctant to use this approach, since I have a lot of keys that then need to be defined twice; both as a type and as the actual message key, but apparently it works to do:
```
import SomeMessage from '.....'

export type SomeMessageKeyType = keyof typeof SomeMessage
```
